crates/sui-framework/sources/governance/staking_pool.move

```diff
@@ -0,0 +1,343 @@
+        validator_address: address,
+        /// The epoch at which the staking pool started operating.
+        pool_starting_epoch: u64,
+        /// The pool tokens representing the amount of rewards the delegator can get back when they withdraw
+        /// from the pool. If this field is `none`, that means the delegation hasn't been activated yet.
+        pool_tokens: Balance<DelegationToken>,
+        /// Number of SUI token staked originally.
+        principal_sui_amount: u64,
+    }
+
+    /// A self-custodial object holding the staked SUI tokens.
+    struct StakedSui has key {
+        id: UID,
+        /// The staked SUI tokens.
+        principal: Balance<SUI>,
+        /// If the stake comes from a Coin<SUI>, this field is None. If it comes from a LockedCoin<SUI>, this
+        /// field will record the original lock expiration epoch, to be used when unstaking.
+        sui_token_lock: Option<EpochTimeLock>,
+    }
+
+    /// Create a new, empty staking pool.
+    public(friend) fun new(validator_address: address, starting_epoch: u64) : StakingPool {
+        StakingPool {
+            validator_address,
+            starting_epoch,
+            epoch_starting_sui_balance: 0,
+            sui_balance: 0,
+            rewards_pool: balance::zero(),
+            delegation_token_supply: balance::create_supply(DelegationToken {}),
+            pending_delegations: vec_map::empty(),
+        }
+    }
+
+    /// Called at epoch advancement times to add rewards (in SUI) to the staking pool, and distribute new delegation tokens. 
+    public(friend) fun advance_epoch(pool: &mut StakingPool, rewards: Balance<SUI>, ctx: &mut TxContext) {
+        pool.sui_balance = pool.sui_balance + balance::value(&rewards);
+        balance::join(&mut pool.rewards_pool, rewards);
+        
+        // distribute pool tokens at new exchange rate.
+        while (!vec_map::is_empty(&pool.pending_delegations)) {
+            let (delegator, sui_amount) = vec_map::pop(&mut pool.pending_delegations);
+            distribute_delegation_tokens(pool, delegator, sui_amount, ctx);
+            pool.sui_balance = pool.sui_balance + sui_amount
+        };
+
+        // Record the epoch starting balance.
+        pool.epoch_starting_sui_balance = pool.sui_balance;
+    }
+
+    /// Request to delegate to a staking pool. The delegation gets counted at the beginning of the next epoch,
+    /// when the delegation object containing the pool tokens is distributed to the delegator.
+    public(friend) fun request_add_delegation(
+        pool: &mut StakingPool, 
+        stake: Balance<SUI>, 
+        sui_token_lock: Option<EpochTimeLock>, 
+        ctx: &mut TxContext
+    ) {
+        let sui_amount = balance::value(&stake);
+        assert!(sui_amount > 0, 0);
+        let delegator = tx_context::sender(ctx);
+        // insert delegation info into the vec map. 
+        let idx_opt = vec_map::get_idx_opt(&pool.pending_delegations, &delegator);
```

Comment:
What's the reason to use VecMap instead of allowing multiple entries from the same delegator?
(slightly concerned about how slow this becomes)

Reply:
Good point. The performance of this function is pretty important. I was thinking about the scenario where one user sends a ton of small delegation requests but I guess we can rate limit the number of new delegation requests per epoch, instead of the number of new delegators. Will change this to vector instead.